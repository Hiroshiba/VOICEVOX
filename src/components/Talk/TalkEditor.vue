<template>
  <span>talk</span>
</template>

<script setup lang="ts">
defineProps<{
  showPitch: boolean;
}>();
</script>
