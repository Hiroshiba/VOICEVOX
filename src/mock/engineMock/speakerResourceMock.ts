/**
 * 話者情報を作るモック。
 * なんとなくENGINEのモックと揃えている。
 */

import { Speaker, SpeakerInfo } from "@/openapi";

/** 話者を返すモック */
export function getSpeakersMock(): Speaker[] {
  return [
    // トーク２つ
    {
      name: "dummy1",
      styles: [
        { name: "style0", id: 0 },
        { name: "style1", id: 2 },
      ],
      speakerUuid: "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
      version: "mock",
    },
    // トーク２つ
    {
      name: "dummy2",
      styles: [
        { name: "style0", id: 1 },
        { name: "style1", id: 3 },
      ],
      speakerUuid: "388f246b-8c41-4ac1-8e2d-5d79f3ff56d9",
      version: "mock",
    },
    // トーク１つ
    {
      name: "dummy3",
      styles: [{ name: "style0", id: 8 }],
      speakerUuid: "35b2c544-660e-401e-b503-0e14c635303a",
      version: "mock",
    },
  ];
}

/** 話者の追加情報を返すモック。 */
export function getSpeakerInfoMock(
  speakerUuid: string,
  assetsDir: string,
): SpeakerInfo {
  const speakers = getSpeakersMock();

  // NOTE: 画像のURLを得るために必要
  const speakerIndex = speakers.findIndex(
    (speaker) => speaker.speakerUuid === speakerUuid,
  );
  if (speakerIndex === -1) {
    throw new Error(`Speaker not found: ${speakerUuid}`);
  }

  const styleIds = speakers[speakerIndex].styles.map((style) => style.id);

  return {
    policy: `Dummy policy for ${speakerUuid}`,
    portrait: `${assetsDir}/portrait_${speakerIndex + 1}.png`,
    styleInfos: styleIds.map((id) => {
      return {
        id,
        icon: `${assetsDir}/icon_${speakerIndex + 1}.png`,
        voiceSamples: [],
      };
    }),
  };
}
