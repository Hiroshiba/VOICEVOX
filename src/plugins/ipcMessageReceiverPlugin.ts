import { Plugin } from "vue";
import { debounce } from "quasar";
import { Store } from "@/store/vuex";
import { AllActions, AllGetters, AllMutations, State } from "@/store/type";
import { IpcRendererOn } from "@/type/ipc";

// これを直接contextBridgeで渡そうとするとクローン不可で怒られるので、使う側に置く
const onReceivedIPCMsg = new Proxy({} as IpcRendererOn, {
  get:
    (_, channel: keyof IpcRendererOn) =>
    (listener: (event: unknown, ...args: unknown[]) => unknown) => {
      window.backend.onReceivedIPCMsg(channel, listener);
    },
});

export const ipcMessageReceiver: Plugin = {
  install: (
    _,
    options: { store: Store<State, AllGetters, AllActions, AllMutations> },
  ) => {
    onReceivedIPCMsg.LOAD_PROJECT_FILE((_, { filePath, confirm } = {}) =>
      options.store.dispatch("LOAD_PROJECT_FILE", { filePath, confirm }),
    );

    onReceivedIPCMsg.DETECT_MAXIMIZED(() =>
      options.store.dispatch("DETECT_MAXIMIZED"),
    );

    onReceivedIPCMsg.DETECT_UNMAXIMIZED(() =>
      options.store.dispatch("DETECT_UNMAXIMIZED"),
    );

    onReceivedIPCMsg.DETECTED_ENGINE_ERROR((_, { engineId }) =>
      options.store.dispatch("DETECTED_ENGINE_ERROR", { engineId }),
    );

    onReceivedIPCMsg.DETECT_PINNED(() => {
      void options.store.dispatch("DETECT_PINNED");
    });

    onReceivedIPCMsg.DETECT_UNPINNED(() => {
      void options.store.dispatch("DETECT_UNPINNED");
    });

    onReceivedIPCMsg.DETECT_ENTER_FULLSCREEN(() =>
      options.store.dispatch("DETECT_ENTER_FULLSCREEN"),
    );

    onReceivedIPCMsg.DETECT_LEAVE_FULLSCREEN(() =>
      options.store.dispatch("DETECT_LEAVE_FULLSCREEN"),
    );

    onReceivedIPCMsg.CHECK_EDITED_AND_NOT_SAVE((_, obj) => {
      void options.store.dispatch("CHECK_EDITED_AND_NOT_SAVE", obj);
    });

    onReceivedIPCMsg.DETECT_RESIZED(
      debounce(
        (_, { width, height }: { width: number; height: number }) =>
          window.dataLayer?.push({ event: "windowResize", width, height }),
        300,
      ),
    );
  },
};
