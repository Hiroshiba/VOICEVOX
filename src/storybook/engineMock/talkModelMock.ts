/**
 * トーク系の構造体を作るモック。
 */

import { IpadicFeatures } from "kuromoji";
import { moraToPhonemes } from "./phonemeMock";
import { convertHiraToKana, moraPattern } from "@/domain/japanese";
import { AccentPhrase, Mora } from "@/openapi";

/** アルファベット文字列を適当な0~1の適当な数値に変換する */
function alphabetsToNumber(text: string): number {
  const codes = text.split("").map((c) => c.charCodeAt(0));
  const sum = codes.reduce((a, b) => a + b);
  return (sum % 256) / 256;
}

// 0.1~1になるように適当な長さを求める
function phonemeToLengthMock(phoneme: string): number {
  return alphabetsToNumber(phoneme) * 0.9 + 0.1;
}

// 3~6になるように適当なピッチを求める
function phonemeToPitchMock(phoneme: string): number {
  return (1 - alphabetsToNumber(phoneme)) * 3 + 3;
}

/** カタカナテキストを適当なモーラに変換する */
function textToMoraMock(text: string): Mora {
  const phonemes = moraToPhonemes[text];
  if (phonemes == undefined) throw new Error(`モーラに変換できません: ${text}`);

  return {
    text,
    consonant: phonemes[0],
    consonantLength:
      phonemes[0] == undefined ? undefined : phonemeToLengthMock(phonemes[0]),
    vowel: phonemes[1],
    vowelLength: phonemeToLengthMock(phonemes[1]),
    pitch: phonemeToPitchMock((phonemes[0] ?? "") + phonemes[1]),
  };
}

/** カタカナテキストを適当なアクセント句に変換する */
function textToAccentPhrase(text: string): AccentPhrase {
  const moras: Mora[] = [...text.matchAll(moraPattern)].map((m) =>
    textToMoraMock(m[0]),
  );
  const alphabets = moras.map((m) => (m.consonant ?? "") + m.vowel).join("");
  const accent =
    1 + Math.round(alphabetsToNumber(alphabets) * (moras.length - 1));
  return { moras, accent };
}

/**
 * テキストを適当なアクセント句に分割する。
 * 助詞ごとに区切る。記号ごとに無音を入れる。
 * 無音で終わるアクセント句の最後のモーラが「す」「つ」の場合は無声化する。
 */
export function tokensToActtentPhrases(tokens: IpadicFeatures[]) {
  // TODO: 先にアクセントを求めてから長さなどを求める
  const accentPhrases: AccentPhrase[] = [];
  let textPhrase = "";
  for (const token of tokens) {
    // 記号の場合は無音を入れて区切る
    if (token.pos == "記号") {
      if (textPhrase.length == 0) continue;

      const accentPhrase = textToAccentPhrase(textPhrase);
      accentPhrase.pauseMora = {
        text: "、",
        vowel: "",
        vowelLength: 1 - 1 / (accentPhrases.length + 1),
        pitch: 0,
      };
      accentPhrases.push(accentPhrase);
      textPhrase = "";
      continue;
    }

    // 記号以外は連結
    if (token.reading == undefined)
      throw new Error(`発音がないトークン: ${token.surface_form}`);
    textPhrase += token.reading;

    // 助詞の場合は区切る
    if (token.pos == "助詞") {
      accentPhrases.push(textToAccentPhrase(textPhrase));
      textPhrase = "";
    }
  }
  if (textPhrase != "") {
    accentPhrases.push(textToAccentPhrase(textPhrase));
  }

  // 最後のアクセント句の無音をなくす
  if (accentPhrases.length > 0) {
    const lastPhrase = accentPhrases[accentPhrases.length - 1];
    lastPhrase.pauseMora = undefined;
  }

  // 無音のあるアクセント句を無声化
  for (const phrase of accentPhrases) {
    if (phrase.pauseMora == undefined) continue;
    const lastMora = phrase.moras[phrase.moras.length - 1];
    if (lastMora.text == "ス" || lastMora.text == "ツ") {
      lastMora.vowel = "U";
      lastMora.pitch = 0;
    }
  }

  // 別のアクセント句で同じにならないように適当に値をずらす
  for (let i = 0; i < accentPhrases.length; i++) {
    const diff = i * 0.01;
    const accentPhrase = accentPhrases[i];
    for (const mora of accentPhrase.moras) {
      if (mora.consonantLength != undefined) mora.consonantLength += diff;
      mora.vowelLength += diff;
      if (mora.pitch > 0) mora.pitch += diff;
    }
    if (accentPhrase.pauseMora != undefined) {
      accentPhrase.pauseMora.vowelLength += diff;
    }
  }

  return accentPhrases;
}
