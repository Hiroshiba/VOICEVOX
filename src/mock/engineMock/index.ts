import kuromoji, { IpadicFeatures, Tokenizer } from "kuromoji";
import { getEngineManifestMock } from "./manifestMock";
import {
  replaseLengthMock,
  replasePitchMock,
  tokensToActtentPhrasesMock,
} from "./talkModelMock";
import { getSpeakerInfoMock, getSpeakersMock } from "./speakerResourceMock";
import { IEngineConnectorFactory } from "@/infrastructures/EngineConnector";
import {
  AccentPhrase,
  AccentPhrasesAccentPhrasesPostRequest,
  AudioQuery,
  AudioQueryAudioQueryPostRequest,
  DefaultApiInterface,
  EngineManifest,
  MoraDataMoraDataPostRequest,
  Speaker,
  SpeakerInfo,
  SpeakerInfoSpeakerInfoGetRequest,
  SupportedDevicesInfo,
} from "@/openapi";

export const dicPath = "engineMock/dict";
export const assetsPath = "engineMock/assets";

export const mockHost = "mock";

export function createOpenAPIEngineMock(): IEngineConnectorFactory {
  let mockApi: Partial<DefaultApiInterface> | undefined = undefined;
  return {
    instance: (host: string) => {
      if (host !== mockHost) {
        throw new Error(`Invalid host: ${host}`);
      }

      if (mockApi == undefined) {
        // テキスト形態素解析器
        const tokenizerPromise = new Promise<Tokenizer<IpadicFeatures>>(
          (resolve, reject) => {
            kuromoji
              .builder({
                dicPath,
              })
              .build(function (err, tokenizer) {
                if (err) {
                  reject(err);
                } else {
                  resolve(tokenizer);
                }
              });
          },
        );

        mockApi = {
          async versionVersionGet(): Promise<string> {
            return "mock";
          },

          async engineManifestEngineManifestGet(): Promise<EngineManifest> {
            return getEngineManifestMock();
          },

          async supportedDevicesSupportedDevicesGet(): Promise<SupportedDevicesInfo> {
            return { cpu: true, cuda: false, dml: false };
          },

          async isInitializedSpeakerIsInitializedSpeakerGet(): Promise<boolean> {
            return true;
          },

          async initializeSpeakerInitializeSpeakerPost(): Promise<void> {
            return;
          },

          async speakersSpeakersGet(): Promise<Speaker[]> {
            return getSpeakersMock();
          },

          async speakerInfoSpeakerInfoGet(
            payload: SpeakerInfoSpeakerInfoGetRequest,
          ): Promise<SpeakerInfo> {
            if (payload.resourceFormat != "url")
              throw new Error("resourceFormatはurl以外未対応です");
            return getSpeakerInfoMock(payload.speakerUuid, assetsPath);
          },

          async audioQueryAudioQueryPost(
            payload: AudioQueryAudioQueryPostRequest,
          ): Promise<AudioQuery> {
            const tokenizer = await tokenizerPromise;
            const tokens = tokenizer.tokenize(payload.text);
            const accentPhrases = tokensToActtentPhrasesMock(
              tokens,
              payload.speaker,
            );

            return {
              accentPhrases,
              speedScale: 1.0,
              pitchScale: 0,
              intonationScale: 1.0,
              volumeScale: 1.0,
              prePhonemeLength: 0.1,
              postPhonemeLength: 0.1,
              outputSamplingRate: 24000,
              outputStereo: false,
            };
          },

          async accentPhrasesAccentPhrasesPost(
            payload: AccentPhrasesAccentPhrasesPostRequest,
          ): Promise<AccentPhrase[]> {
            if (payload.isKana == true)
              throw new Error("AquesTalk風記法は未対応です");

            const tokenizer = await tokenizerPromise;
            const tokens = tokenizer.tokenize(payload.text);
            const accentPhrases = tokensToActtentPhrasesMock(
              tokens,
              payload.speaker,
            );
            return accentPhrases;
          },

          async moraDataMoraDataPost(
            payload: MoraDataMoraDataPostRequest,
          ): Promise<AccentPhrase[]> {
            replaseLengthMock(payload.accentPhrase, payload.speaker);
            replasePitchMock(payload.accentPhrase, payload.speaker);
            return payload.accentPhrase;
          },
        };
      }

      return mockApi as DefaultApiInterface;
    },
  };
}
