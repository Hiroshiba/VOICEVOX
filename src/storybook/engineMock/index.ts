import kuromoji, { IpadicFeatures, Tokenizer } from "kuromoji";
import { tokensToActtentPhrases } from "./talkModelMock";
import { IEngineConnectorFactory } from "@/infrastructures/EngineConnector";
import {
  AccentPhrase,
  AccentPhrasesAccentPhrasesPostRequest,
  DefaultApiInterface,
  Mora,
} from "@/openapi";
import { moraPattern } from "@/domain/japanese";

export const dicPath = "/kuromoji/dict";

export const mockHost = "mock";

export function createOpenAPIEngineMock(): IEngineConnectorFactory {
  let mockApi: Partial<DefaultApiInterface> | undefined = undefined;
  return {
    instance: (host: string) => {
      if (host !== mockHost) {
        throw new Error(`Invalid host: ${host}`);
      }

      if (mockApi == undefined) {
        // テキスト形態素解析器
        const tokenizerPromise = new Promise<Tokenizer<IpadicFeatures>>(
          (resolve, reject) => {
            kuromoji
              .builder({
                dicPath,
              })
              .build(function (err, tokenizer) {
                if (err) {
                  reject(err);
                } else {
                  resolve(tokenizer);
                }
              });
          },
        );

        mockApi = {
          async accentPhrasesAccentPhrasesPost(
            payload: AccentPhrasesAccentPhrasesPostRequest,
          ): Promise<AccentPhrase[]> {
            if (payload.isKana == true)
              throw new Error("AquesTalk風記法は未対応です");

            const tokenizer = await tokenizerPromise;
            const tokens = tokenizer.tokenize(payload.text);
            const accentPhrases = tokensToActtentPhrases(tokens);

            // 話者ごとに同じにならないように適当にずらす
            const speaker = payload.speaker;
            const diff = speaker * 0.03;
            for (const accentPhrase of accentPhrases) {
              for (const mora of accentPhrase.moras) {
                if (mora.consonantLength != undefined)
                  mora.consonantLength += diff;
                mora.vowelLength += diff;
                if (mora.pitch > 0) mora.pitch += diff;
              }
              if (accentPhrase.pauseMora != undefined) {
                accentPhrase.pauseMora.vowelLength += diff;
              }
            }

            return accentPhrases;
          },
        };
      }

      return mockApi as DefaultApiInterface;
    },
  };
}
