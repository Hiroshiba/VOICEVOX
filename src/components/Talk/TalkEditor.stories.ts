import { userEvent, within, expect, fn } from "@storybook/test";

import { Meta, StoryObj } from "@storybook/vue3";
import { provide, toRaw } from "vue";
import TalkEditor from "./TalkEditor.vue";
import { createStoreWrapper, storeKey } from "@/store";
import { HotkeyManager, hotkeyManagerKey } from "@/plugins/hotkeyPlugin";
import { StoreType } from "@/store/type";
import {
  assetsPath,
  createOpenAPIEngineMock,
  mockHost,
} from "@/storybook/engineMock";
import { proxyStoreCreator } from "@/store/proxy";
import { CharacterInfo, EngineId, SpeakerId, StyleId } from "@/type/preload";
import { getEngineManifestMock } from "@/storybook/engineMock/manifestMock";
import {
  getSpeakerInfoMock,
  getSpeakersMock,
} from "@/storybook/engineMock/speakerResourceMock";

const meta: Meta<typeof TalkEditor> = {
  component: TalkEditor,
  args: {
    isEnginesReady: true,
    isProjectFileLoaded: false,
  },
  decorators: [
    (story, context) => {
      const hotkeyManager = new HotkeyManager();
      provide(hotkeyManagerKey, hotkeyManager);

      const store = createStoreWrapper({
        proxyStoreDI: proxyStoreCreator(createOpenAPIEngineMock()),
      });
      store.dispatch("HYDRATE_SETTING_STORE"); // FIXME: 色設定取得のため。設定も読み込んでしまうため不要にしたい。

      // context.parameters.store = store;

      const engineManifest = getEngineManifestMock();
      const engineId = EngineId(engineManifest.uuid);

      // setup store
      const speakers = getSpeakersMock();
      const characterInfos: CharacterInfo[] = speakers.map((speaker) => {
        const speakerInfo = getSpeakerInfoMock(speaker.speakerUuid, assetsPath);
        return {
          portraitPath: speakerInfo.portrait,
          metas: {
            speakerUuid: SpeakerId(speaker.speakerUuid),
            speakerName: speaker.name,
            styles: speakerInfo.styleInfos.map((styleInfo) => {
              const style = speaker.styles.find((s) => s.id === styleInfo.id);
              if (style == undefined) throw new Error("style not found");
              return {
                styleName: style.name,
                styleId: StyleId(style.id),
                styleType: style.type,
                iconPath: styleInfo.icon,
                portraitPath: styleInfo.portrait ?? speakerInfo.portrait,
                engineId,
                voiceSamplePaths: styleInfo.voiceSamples,
              };
            }),
            policy: speakerInfo.policy,
          },
        };
      });

      store.replaceState({
        ...structuredClone(toRaw(store.state)),
        engineIds: [engineId],
        engineStates: {
          [engineId]: "READY",
        },
        engineInfos: {
          [engineId]: {
            uuid: engineId,
            host: mockHost,
            name: engineManifest.name,
            path: undefined,
            executionEnabled: false,
            executionFilePath: "not_found",
            executionArgs: [],
            type: "default",
          },
        },
        engineManifests: {
          [engineId]: engineManifest,
        },
        characterInfos: { [engineId]: characterInfos },
        defaultStyleIds: speakers.map((speaker) => ({
          engineId: engineId,
          speakerUuid: SpeakerId(speaker.speakerUuid),
          defaultStyleId: StyleId(speaker.styles[0].id),
        })),
      });

      provide(storeKey, store);

      return story();
    },
  ],
  beforeEach: async ({ parameters }) => {
    // const store = parameters.store; // TODO: 型を付けたい
    // await store.dispatch("LOAD_CHARACTER", { engineId: EngineId(mockHost) });
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  name: "デフォルト",
};

export const NowLoading: Story = {
  name: "プロジェクトファイルを読み込み中",
  args: {
    isProjectFileLoaded: "waiting",
  },
};
