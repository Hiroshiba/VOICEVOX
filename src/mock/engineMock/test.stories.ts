import { userEvent, within, expect, fn } from "@storybook/test";

import { Meta, StoryObj } from "@storybook/vue3";
import { createOpenAPIEngineMock, mockHost } from "@/mock/engineMock";

const meta: Meta = {
  render: () => {
    return { template: "<div/>" };
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const AudioQuery: Story = {
  play: async ({ args }) => {
    const mockApi = createOpenAPIEngineMock();
    const api = mockApi.instance(mockHost);
    const result = await api.audioQueryAudioQueryPost({
      text: "そうです、そのとおりだと思いますね",
      speaker: 1,
    });
    console.log(result);
  },
};

export const Synthesis: Story = {
  play: async ({ args }) => {
    const mockApi = createOpenAPIEngineMock();
    const api = mockApi.instance(mockHost);
    const audioQuery = await api.audioQueryAudioQueryPost({
      text: "そうです、そのとおりだと思いますね",
      speaker: 1,
    });
    const blob = await api.synthesisSynthesisPost({
      audioQuery,
      speaker: 1,
    });
    console.log(blob);
  },
};
