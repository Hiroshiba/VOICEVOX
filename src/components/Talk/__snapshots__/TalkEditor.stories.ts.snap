// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Talk/TalkEditor TextInput play-test 1`] = `
{
  "audioItems": {
    "00000000-0000-4000-0000-000000000001": {
      "presetKey": undefined,
      "query": {
        "accentPhrases": [
          {
            "accent": 5,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0220625,
                "pitch": 4.7734375,
                "text": "コ",
                "vowel": "o",
                "vowelLength": 0.1140625,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "ン",
                "vowel": "N",
                "vowelLength": 0.08312499999999999,
              },
              {
                "consonant": "n",
                "consonantLength": 0.022625,
                "pitch": 4.75,
                "text": "ニ",
                "vowel": "i",
                "vowelLength": 0.10843749999999999,
              },
              {
                "consonant": "ch",
                "consonantLength": 0.0400625,
                "pitch": 4.0234375,
                "text": "チ",
                "vowel": "i",
                "vowelLength": 0.10843749999999999,
              },
              {
                "consonant": "h",
                "consonantLength": 0.0315,
                "pitch": 4.796875,
                "text": "ハ",
                "vowel": "a",
                "vowelLength": 0.1509375,
              },
            ],
            "pauseMora": {
              "pitch": 0,
              "text": "、",
              "vowel": "pau",
              "vowelLength": 0,
            },
          },
          {
            "accent": 2,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0320625,
                "pitch": 4.7834375,
                "text": "コ",
                "vowel": "o",
                "vowelLength": 0.12406249999999999,
              },
              {
                "consonant": "r",
                "consonantLength": 0.033375,
                "pitch": 4.72875,
                "text": "レ",
                "vowel": "e",
                "vowelLength": 0.11468749999999998,
              },
              {
                "consonant": "h",
                "consonantLength": 0.0415,
                "pitch": 5.106875,
                "text": "ハ",
                "vowel": "a",
                "vowelLength": 0.1609375,
              },
            ],
          },
          {
            "accent": 2,
            "moras": [
              {
                "consonant": "t",
                "consonantLength": 0.04375,
                "pitch": 4.723125,
                "text": "テ",
                "vowel": "e",
                "vowelLength": 0.12468749999999999,
              },
              {
                "consonant": "s",
                "consonantLength": 0.0435625,
                "pitch": 4.7309375,
                "text": "ス",
                "vowel": "u",
                "vowelLength": 0.1396875,
              },
              {
                "consonant": "t",
                "consonantLength": 0.04375,
                "pitch": 5.023124999999999,
                "text": "ト",
                "vowel": "o",
                "vowelLength": 0.1340625,
              },
              {
                "consonant": "d",
                "consonantLength": 0.040749999999999995,
                "pitch": 4.848125,
                "text": "デ",
                "vowel": "e",
                "vowelLength": 0.12468749999999999,
              },
              {
                "consonant": "s",
                "consonantLength": 0.0535625,
                "pitch": 4.7309375,
                "text": "ス",
                "vowel": "u",
                "vowelLength": 0.18968749999999998,
              },
            ],
            "pauseMora": undefined,
          },
        ],
        "intonationScale": 1,
        "outputSamplingRate": 8000,
        "outputStereo": false,
        "pitchScale": 0,
        "postPhonemeLength": 0.1,
        "prePhonemeLength": 0.1,
        "speedScale": 1,
        "volumeScale": 1,
      },
      "text": "こんにちは、これはテストです。",
      "voice": {
        "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
        "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
        "styleId": 0,
      },
    },
  },
  "audioKeys": [
    "00000000-0000-4000-0000-000000000001",
  ],
}
`;

exports[`components/Talk/TalkEditor TextPaste play-test 1`] = `
{
  "audioItems": {
    "00000000-0000-4000-0000-000000000001": {
      "presetKey": undefined,
      "query": {
        "accentPhrases": [
          {
            "accent": 4,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0220625,
                "pitch": 4.7734375,
                "text": "カ",
                "vowel": "a",
                "vowelLength": 0.10093749999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.10843749999999999,
              },
              {
                "consonant": "gy",
                "consonantLength": 0.044,
                "pitch": 3.859375,
                "text": "ギョ",
                "vowel": "o",
                "vowelLength": 0.1140625,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.11968749999999999,
              },
              {
                "consonant": "d",
                "consonantLength": 0.03075,
                "pitch": 5.128125,
                "text": "デ",
                "vowel": "e",
                "vowelLength": 0.15468749999999998,
              },
            ],
          },
          {
            "accent": 4,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0320625,
                "pitch": 4.7834375,
                "text": "カ",
                "vowel": "a",
                "vowelLength": 0.11093749999999998,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.11843749999999999,
              },
              {
                "consonant": "gy",
                "consonantLength": 0.054,
                "pitch": 3.869375,
                "text": "ギョ",
                "vowel": "o",
                "vowelLength": 0.12406249999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.1796875,
              },
            ],
            "pauseMora": undefined,
          },
        ],
        "intonationScale": 1,
        "outputSamplingRate": 8000,
        "outputStereo": false,
        "pitchScale": 0,
        "postPhonemeLength": 0.1,
        "prePhonemeLength": 0.1,
        "speedScale": 1,
        "volumeScale": 1,
      },
      "text": "改行で改行",
      "voice": {
        "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
        "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
        "styleId": 0,
      },
    },
    "00000000-0000-4000-0000-000000000003": {
      "morphingInfo": undefined,
      "presetKey": undefined,
      "query": {
        "accentPhrases": [
          {
            "accent": 2,
            "moras": [
              {
                "consonant": "t",
                "consonantLength": 0.02375,
                "pitch": 4.703125,
                "text": "ト",
                "vowel": "o",
                "vowelLength": 0.1140625,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.11968749999999999,
              },
              {
                "consonant": "t",
                "consonantLength": 0.02375,
                "pitch": 5.003125,
                "text": "テ",
                "vowel": "e",
                "vowelLength": 0.10468749999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "ン",
                "vowel": "N",
                "vowelLength": 0.08312499999999999,
              },
              {
                "consonant": "d",
                "consonantLength": 0.03075,
                "pitch": 4.828125,
                "text": "デ",
                "vowel": "e",
                "vowelLength": 0.15468749999999998,
              },
            ],
          },
          {
            "accent": 4,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0320625,
                "pitch": 4.7834375,
                "text": "カ",
                "vowel": "a",
                "vowelLength": 0.11093749999999998,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.11843749999999999,
              },
              {
                "consonant": "gy",
                "consonantLength": 0.054,
                "pitch": 3.869375,
                "text": "ギョ",
                "vowel": "o",
                "vowelLength": 0.12406249999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.1796875,
              },
            ],
            "pauseMora": undefined,
          },
        ],
        "intonationScale": 1,
        "outputSamplingRate": 8000,
        "outputStereo": false,
        "pitchScale": 0,
        "postPhonemeLength": 0.1,
        "prePhonemeLength": 0.1,
        "speedScale": 1,
        "volumeScale": 1,
      },
      "text": "読点で改行。",
      "voice": {
        "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
        "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
        "styleId": 0,
      },
    },
    "00000000-0000-4000-0000-000000000004": {
      "morphingInfo": undefined,
      "presetKey": undefined,
      "query": {
        "accentPhrases": [
          {
            "accent": 4,
            "moras": [
              {
                "consonant": "s",
                "consonantLength": 0.023562499999999997,
                "pitch": 4.7109375,
                "text": "サ",
                "vowel": "a",
                "vowelLength": 0.10093749999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.609375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.10843749999999999,
              },
              {
                "consonant": "g",
                "consonantLength": 0.021312499999999998,
                "pitch": 4.8046875,
                "text": "ゴ",
                "vowel": "o",
                "vowelLength": 0.1140625,
              },
              {
                "consonant": "n",
                "consonantLength": 0.032625,
                "pitch": 4.75,
                "text": "ノ",
                "vowel": "o",
                "vowelLength": 0.1640625,
              },
            ],
          },
          {
            "accent": 2,
            "moras": [
              {
                "consonant": "t",
                "consonantLength": 0.03375,
                "pitch": 4.713125,
                "text": "ト",
                "vowel": "o",
                "vowelLength": 0.12406249999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.12968749999999998,
              },
              {
                "consonant": "t",
                "consonantLength": 0.03375,
                "pitch": 5.013125,
                "text": "テ",
                "vowel": "e",
                "vowelLength": 0.11468749999999998,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.619375,
                "text": "ン",
                "vowel": "N",
                "vowelLength": 0.09312499999999999,
              },
              {
                "consonant": "h",
                "consonantLength": 0.0415,
                "pitch": 4.806875,
                "text": "ハ",
                "vowel": "a",
                "vowelLength": 0.1609375,
              },
            ],
          },
          {
            "accent": 4,
            "moras": [
              {
                "consonant": "k",
                "consonantLength": 0.0420625,
                "pitch": 4.7934375,
                "text": "カ",
                "vowel": "a",
                "vowelLength": 0.12093749999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.629375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.12843749999999998,
              },
              {
                "consonant": "gy",
                "consonantLength": 0.064,
                "pitch": 3.879375,
                "text": "ギョ",
                "vowel": "o",
                "vowelLength": 0.1340625,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.629375,
                "text": "ウ",
                "vowel": "u",
                "vowelLength": 0.1396875,
              },
              {
                "consonant": "sh",
                "consonantLength": 0.0630625,
                "pitch": 4.218437499999999,
                "text": "シ",
                "vowel": "i",
                "vowelLength": 0.12843749999999998,
              },
              {
                "consonant": "n",
                "consonantLength": 0.042624999999999996,
                "pitch": 4.77,
                "text": "ナ",
                "vowel": "a",
                "vowelLength": 0.12093749999999999,
              },
              {
                "consonant": undefined,
                "consonantLength": undefined,
                "pitch": 3.629375,
                "text": "イ",
                "vowel": "i",
                "vowelLength": 0.1784375,
              },
            ],
            "pauseMora": undefined,
          },
        ],
        "intonationScale": 1,
        "outputSamplingRate": 8000,
        "outputStereo": false,
        "pitchScale": 0,
        "postPhonemeLength": 0.1,
        "prePhonemeLength": 0.1,
        "speedScale": 1,
        "volumeScale": 1,
      },
      "text": "最後の読点は改行しない。",
      "voice": {
        "engineId": "c7b58856-bd56-4aa1-afb7-b8415f824b06",
        "speakerId": "7ffcb7ce-00ec-4bdc-82cd-45a8889e43ff",
        "styleId": 0,
      },
    },
  },
  "audioKeys": [
    "00000000-0000-4000-0000-000000000001",
    "00000000-0000-4000-0000-000000000003",
    "00000000-0000-4000-0000-000000000004",
  ],
}
`;
